### PR TITLE
Elo ranking base

### PR DESCRIPTION
**Pull request by AkkarinForest**

### CODE REVIEW

File: app/services/elo/next_ranking_calculator.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module Elo
+  class NextRankingCalculator
+    def initialize(points_change_calculator: PointsChangeCalculator.new)
+      @points_change_calculator = points_change_calculator
+    end
+
+    def call(current_ranking, match)
+      team_a_points = players_average_points(current_ranking, *match.players_from_team(Match::TEAM_1))
```
Comment (shadre) [inline app/services/elo/next_ranking_calculator.rb]: 👍

File: app/services/elo/next_ranking_calculator.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+      next_ranking(current_ranking, players_points_update)
+    end
+
+    private
+
+    attr :points_change_calculator
+
+    def players_average_points(current_ranking, first_player, second_player)
+      (current_ranking[first_player] + current_ranking[second_player]) / 2
```
Comment (shadre) [inline app/services/elo/next_ranking_calculator.rb]: Are those two numbers always floats? If not, or not sure, I would change `2` to `2.0`.

File: app/services/elo/next_ranking_calculator.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    end
+
+    def next_ranking(current_ranking, players_points_update)
+      players_points_update
+      .each
```
Comment (shadre) [inline app/services/elo/next_ranking_calculator.rb]: I would indent 2 spaces here

File: app/services/elo/next_ranking_calculator.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      .with_object(current_ranking.dup) do |(player, points), next_ranking|
+        next_ranking[player] += points
+      end
+    end
+
+    def extract_players_points_update(match, teams_points_update)
```
Comment (shadre) [inline app/services/elo/next_ranking_calculator.rb]: I would work on naming here a little bit. Not sure "extracting" is what's going on here.

File: app/services/elo/ranking_calculator.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module Elo
+  class RankingCalculator
+    def initialize(next_ranking_calculator: NextRankingCalculator.new)
+      @next_ranking_calculator = next_ranking_calculator
+    end
+
+    def call
+      matches.reduce(initial_ranking) do | current_ranking, match |
```
Comment (shadre) [inline app/services/elo/ranking_calculator.rb]: There are unneeded spaces between pipes (`|`)

File: config/routes.rb [hunk 6]
```diff
@@ -2,12 +2,13 @@
...
-  
+
   namespace :api do
     resources :matches, only: :create
   end
   resources :matches, only: :index
 
   resources :players, only: [:index, :show]
   get :ranking, action: :ranking, controller: :players
+  get 'ranking/elo', action: :show, controller: :ranking
```
Comment (shadre) [inline config/routes.rb]: I'm unsure about this division between two controllers (lines 12-13).
So, the `Players` controller currently responds to a request at the `/ranking/` path, and the `Ranking` controller is responsible for `/ranking/elo`.
It does seem a little confusing to me.

File: spec/services/elo/next_ranking_calculator_spec.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  let(:points_change_calc) { instance_double("points_change_calculator") }
+  subject { described_class.new(points_change_calculator: points_change_calc) }
+
+  let!(:mark) { create(:player) }
+  let!(:anna) { create(:player) }
+  let!(:joe) { create(:player) }
+  let!(:peter) { create(:player) }
+  let(:current_ranking) { {mark => 1000, anna => 1000, joe => 1000, peter => 1000} }
+
+#TODO it doesn't check if team_a_points are counted correctly
```
Comment (shadre) [inline spec/services/elo/next_ranking_calculator_spec.rb]: I don't get this comment. Should this spec ever _care_ if the points are counted correctly, or should it be a sole responsibility of a spec testing points change calculator?

File: config/routes.rb (hunk 6 above)
Comment (AkkarinForest) [reply config/routes.rb]: I definitely should be only temporary. We have two choices here: everything is managed by players controllers, or both rankings are managed by the ranking controller and the players page and statistics are managed by players controller. I would vote for the second option. So I would live this as it is now and have a plan to move the old ranking to the ranking controller in the future.

File: spec/services/elo/next_ranking_calculator_spec.rb (hunk 7 above)
Comment (AkkarinForest) [reply spec/services/elo/next_ranking_calculator_spec.rb]: Oh, sure, most of this is counted in points change calc, but I meant that this one count the average points for a team, and this math is not tested anywhere. I think I have a way to fix it though :)

File: app/services/elo/ranking_calculator.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    private
+
+    attr :next_ranking_calculator
+
+    def matches
+      Match.all
+    end
+
+    def initial_ranking
+      Player.all.map { |player| [player, 1000.0] }.to_h
```
Comment (shadre) [inline app/services/elo/ranking_calculator.rb]: It would be better if this `1000.0` number was declared in a constant in my opinion.

**Pull request merged**